File: frontend/src/stores/product.ts
import { defineStore } from 'pinia'
import {
  getProductsApi,
  getProductByIDApi,
  createProductApi,
  updateProductApi,
  deleteProductApi,
  getProductCategoriesApi,
  createProductCategoryApi,
  updateProductCategoryApi,
  deleteProductCategoryApi,
  getWarehousesApi,
  createWarehouseApi,
  updateWarehouseApi,
  deleteWarehouseApi,
  createStockMovementApi,
  getProductMovementsApi,
} from '@/api/product'

export interface ProductCategory {
  id: string
  company_id: string
  name: string
  default_kdv_rate: string
}

export interface Warehouse {
  id: string
  company_id: string
  name: string
  address: string
  is_default: boolean
}

export interface Product {
  id: string
  company_id: string
  code: string
  name: string
  type: 'product' | 'service'
  unit: string
  barcode: string
  custom_codes: string
  brand?: string
  serial_numbers: string
  purchase_price: string
  average_cost: string
  sale_price: string
  currency: string
  tax_included: boolean
  purchase_tax_included: boolean
  tax_rate: string
  purchase_tax_rate: string
  description: string
  track_stock: boolean
  current_stock: string
  min_stock: string
  category_id: string | null
  category?: ProductCategory
  is_active: boolean
  created_at: string
}

export interface StockMovement {
  id: string
  company_id: string
  product_id: string
  product?: Product
  warehouse_id: string
  warehouse?: Warehouse
  date: string
  type: 'in' | 'out'
  source_type: 'invoice' | 'manual' | 'transfer'
  source_id: string | null
  quantity: string
  unit_cost: string
  balance_after: string
  note: string
  created_at: string
}

export const useProductStore = defineStore('product', {
  state: () => ({
    products: [] as Product[],
    total: 0,
    categories: [] as ProductCategory[],
    warehouses: [] as Warehouse[],
    movements: [] as StockMovement[],
    loading: false,
  }),
  actions: {
    async fetchProducts(params?: any) {
      this.loading = true
      try {
        const res = await getProductsApi(params)
        this.products = res.data.data
        this.total = res.data.meta.total
      } finally {
        this.loading = false
      }
    },
    async fetchCategories() {
      const res = await getProductCategoriesApi()
      this.categories = res.data.data
    },
    async fetchWarehouses() {
      const res = await getWarehousesApi()
      this.warehouses = res.data.data
    },
    async fetchProductMovements(productId: string) {
      this.loading = true
      try {
        const res = await getProductMovementsApi(productId)
        this.movements = res.data.data
      } finally {
        this.loading = false
      }
    },
    async getProductByID(id: string) {
      const res = await getProductByIDApi(id)
      return res.data.data
    },
    async createProduct(data: any) {
      const res = await createProductApi(data)
      return res.data.data
    },
    async updateProduct(id: string, data: any) {
      const res = await updateProductApi(id, data)
      return res.data.data
    },
    async deleteProduct(id: string) {
      await deleteProductApi(id)
    },
    async createCategory(data: any) {
      const res = await createProductCategoryApi(data)
      await this.fetchCategories()
      return res.data.data
    },
    async updateCategory(id: string, data: any) {
      const res = await updateProductCategoryApi(id, data)
      await this.fetchCategories()
      return res.data.data
    },
    async deleteCategory(id: string) {
      await deleteProductCategoryApi(id)
      await this.fetchCategories()
    },
    async createWarehouse(data: any) {
      const res = await createWarehouseApi(data)
      await this.fetchWarehouses()
      return res.data.data
    },
    async updateWarehouse(id: string, data: any) {
      const res = await updateWarehouseApi(id, data)
      await this.fetchWarehouses()
      return res.data.data
    },
    async deleteWarehouse(id: string) {
      await deleteWarehouseApi(id)
      await this.fetchWarehouses()
    },
    async createStockMovement(data: any) {
      const res = await createStockMovementApi(data)
      return res.data.data
    },
  },
})

File: backend/models/product.go
package models

import (
	"github.com/shopspring/decimal"

	"time"

	"github.com/google/uuid"
)

type ProductCategory struct {
	ID             uuid.UUID       `gorm:"type:uuid;primaryKey" json:"id"`
	CompanyID      uuid.UUID       `gorm:"type:uuid;not null;index" json:"company_id"`
	Name           string          `gorm:"type:varchar(255);not null" json:"name"`
	DefaultKDVRate decimal.Decimal `gorm:"type:numeric(5,2);not null;default:20" json:"default_kdv_rate"`
	CreatedBy      *uuid.UUID      `gorm:"type:uuid" json:"created_by"`
	CreatedAt      time.Time       `json:"created_at"`
	UpdatedAt time.Time  `json:"updated_at"`
}

type Product struct {
	ID            uuid.UUID        `gorm:"type:uuid;primaryKey" json:"id"`
	CompanyID     uuid.UUID        `gorm:"type:uuid;not null;index" json:"company_id"`
	Code          string           `gorm:"type:varchar(100);not null" json:"code"` // unique per company sequence-generated
	Name          string           `gorm:"type:varchar(255);not null" json:"name"`
	Brand         string           `gorm:"type:varchar(255)" json:"brand"`
	Type          string           `gorm:"type:varchar(50);not null;default:'product'" json:"type"` // 'product', 'service'
	Unit          string           `gorm:"type:varchar(50)" json:"unit"`                            // Adet, Kg, etc.
	Barcode       string           `gorm:"type:varchar(100)" json:"barcode"`
	CustomCodes   string           `gorm:"type:varchar(500)" json:"custom_codes"`
	SerialNumbers string           `gorm:"type:text" json:"serial_numbers"`
	PurchasePrice decimal.Decimal  `gorm:"type:numeric(65,4);not null;default:0" json:"purchase_price"`
	AverageCost   decimal.Decimal  `gorm:"type:numeric(65,4);not null;default:0" json:"average_cost"`
	SalePrice     decimal.Decimal  `gorm:"type:numeric(65,4);not null;default:0" json:"sale_price"`
	Currency      string           `gorm:"type:varchar(10);not null;default:'TRY'" json:"currency"`
	TaxIncluded   bool             `gorm:"not null;default:false" json:"tax_included"`          // satış fiyatı KDV dahil mi
	PurchaseTaxIncluded bool       `gorm:"not null;default:false" json:"purchase_tax_included"` // alış fiyatı KDV dahil mi
	TaxRate       decimal.Decimal  `gorm:"type:numeric(5,2);not null;default:20" json:"tax_rate"`          // satış KDV oranı, e.g. 20 for 20%
	PurchaseTaxRate decimal.Decimal `gorm:"type:numeric(5,2);not null;default:20" json:"purchase_tax_rate"` // alış KDV oranı
	Description   string           `gorm:"type:varchar(500)" json:"description"`
	TrackStock    bool             `gorm:"not null;default:true" json:"track_stock"`
	CurrentStock  decimal.Decimal  `gorm:"type:numeric(65,4);not null;default:0" json:"current_stock"` // cached stock quantity
	MinStock      decimal.Decimal  `gorm:"type:numeric(65,4);not null;default:5" json:"min_stock"`
	CategoryID    *uuid.UUID       `gorm:"type:uuid;index" json:"category_id"`
	Category      *ProductCategory `gorm:"foreignKey:CategoryID" json:"category,omitempty"`
	IsActive      bool             `gorm:"not null;default:true" json:"is_active"`
	CreatedBy     *uuid.UUID       `gorm:"type:uuid" json:"created_by"`
	CreatedAt     time.Time        `json:"created_at"`
	UpdatedAt     time.Time        `json:"updated_at"`
}

type Warehouse struct {
	ID        uuid.UUID  `gorm:"type:uuid;primaryKey" json:"id"`
	CompanyID uuid.UUID  `gorm:"type:uuid;not null;index" json:"company_id"`
	Name      string     `gorm:"type:varchar(255);not null" json:"name"`
	Address   string     `gorm:"type:text" json:"address"`
	IsDefault bool       `gorm:"not null;default:false" json:"is_default"`
	CreatedBy *uuid.UUID `gorm:"type:uuid" json:"created_by"`
	CreatedAt time.Time  `json:"created_at"`
	UpdatedAt time.Time  `json:"updated_at"`
}

type StockMovement struct {
	ID           uuid.UUID       `gorm:"type:uuid;primaryKey" json:"id"`
	CompanyID    uuid.UUID       `gorm:"type:uuid;not null;index" json:"company_id"`
	ProductID    uuid.UUID       `gorm:"type:uuid;not null;index" json:"product_id"`
	Product      *Product        `gorm:"foreignKey:ProductID" json:"product,omitempty"`
	WarehouseID  uuid.UUID       `gorm:"type:uuid;not null;index" json:"warehouse_id"`
	Warehouse    *Warehouse      `gorm:"foreignKey:WarehouseID" json:"warehouse,omitempty"`
	Date         time.Time       `gorm:"not null" json:"date"`
	Type         string          `gorm:"type:varchar(20);not null" json:"type"`        // 'in', 'out', 'transfer', 'adjustment'
	SourceType   string          `gorm:"type:varchar(50);not null" json:"source_type"` // 'invoice', 'manual', 'transfer'
	SourceID     *uuid.UUID      `gorm:"type:uuid" json:"source_id"`
	Quantity     decimal.Decimal `gorm:"type:numeric(65,4);not null" json:"quantity"`
	UnitCost     decimal.Decimal `gorm:"type:numeric(65,4);not null;default:0" json:"unit_cost"`
	BalanceAfter decimal.Decimal `gorm:"type:numeric(65,4);not null" json:"balance_after"`
	Note         string          `gorm:"type:varchar(255)" json:"note"`
	CreatedBy    *uuid.UUID      `gorm:"type:uuid" json:"created_by"`
	CreatedAt    time.Time       `json:"created_at"`
	UpdatedAt    time.Time       `json:"updated_at"`
}
